### DISCUSSION TITLE
Custom Modules and Setup

### DISCUSSION THREAD

**Discussion by johndcollins** [OrchardCMS/OrchardCore#11120] [Q&A] [answered]
I have a special requirement for Features etc. so I'm starting from scratch building my own modules. I'm at the early stages and I noticed that the Setup Service hardcodes some values into the database and I'm wondering why it's doing this? My modules are not named the same and I'll not be using these modules. What impact will this have on my application?

The following code is in the SetupService:

![Screenshot 2022-02-03 152657](https://user-images.githubusercontent.com/1684130/152446286-84f43cd7-e926-4bea-8a78-a413785bc64c.png)

**Comment by johndcollins** [comment]
I guess I'll have to make my own Setup service. The current one uses these hardcoded values to register the services a little later and my own recipe service is not used.

It errors out here:
var recipeExecutor = shellContext.ServiceProvider.GetRequiredService<IRecipeExecutor>()

because it's expecting the OrchardCore.Recipe service which it auto registered.
Would it make sense to be able to pass in the required services instead of hard coding them? That would make it more flexible to other uses.

**Comment by ns8482e** [comment, marked as answer]
Setup service is used to initialize the tenant. And to initialize and bootstrap a new tenant on first request orchard core uses special mechanism that depends on few modules that needs to be enabled to execute setup recipes that's you hardcoded. 

However, once the tenant is initialized it will only enable and/or disable modules that you have defined in your setup recipe or in tenant profile. 

Being said - you can use same setup screen However you can define your own setup recipe in your module according to your needs

**Reply by johndcollins** [reply]
Thanks again for you reply. I'm very new to this, just about 3 days in, so I'm still learning.
I'm just surprised that this would be hardcoded. From what I understand, if I wanted my own Feature and Recipe modules, this would not be possible unless I name them Orchard.Recipe and Orchard.Features or am I missing something?

**Reply by ns8482e** [reply]
You can create your own recipe, eg blog recipe here https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Themes/TheBlogTheme/Recipes/blog.recipe.json

Setup screen is just bootstrap mechanism and asks minimum information needed to initialize a tenant. However it's really initialize the tenant by executing recipe you select .

IMO you don't need to override setup screens- instead you build a setup recipe and put it under `Recipes` folder of your module and it'll be available on setup screen to select.
